### ISSUE TITLE
subdirectory file globbing always appends name (creating very long file names), skips top-level files

### ISSUE THREAD

**Issue by jtmoon79** [logrotate/logrotate#474]
**tl;dr** given `logrotate.conf` glob path declaration for `/tmp/path/*/*`, subdirectory files will always get appended `.1`, e.g. `/tmp/path/subA/log.1` becomes `/tmp/path/subA/log.1.1`. Results in ridiculous names like `/tmp/path/subA/log.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1.1`. Files at top-level are skipped.

### Problems

Path declarations glob patterns `/path/*/*` or `/path/**/*` or `/path/*/**` result in two problems:
- errant renaming (append only)
- file matching (failure to rotate log files at top-level)

A brief review of the problem sequence:

1. First pass of `logrotate`
   - file `/path/subA/log` is renamed `/path/subA/log.1`
   - new file `/path/subA/log` created.
   - file `/path/log` is not rotated **_(expected rotation)_**.
2. Second pass of `logrotate`
   - file `/path/subA/log.1` is renamed `/path/subA/log.1.1` **_(expected `/path/subA/log.2`)_**
   - file `/path/subA/log` is renamed `/path/subA/log.1`
   - new file `/path/subA/log` created.
   - file `/path/log` is not rotated **_(expected rotation)_**.

### Reproduction Steps

1. Create test directory `/tmp/path` with dummy log files.
```text
mkdir -vp /tmp/path/subA
dd if=/dev/zero of=/tmp/path/log count=1
dd if=/dev/zero of=/tmp/path/subA/log count=1
```
2. Create minimal `logrotate.conf`
```text
echo "\
create

/tmp/path/**/*
{
        rotate 4
        size 1
}
" > /tmp/logrotate.conf
```
3. review `/tmp/path` contents before first pass
```text
$ find /tmp/path -type f -printf '%s\t%p\n' | sort -k2
512     /tmp/path/log
512     /tmp/path/subA/log
```

##### First pass

4. run `logrotate` for the first pass, (remove repetitive message `Creating new state` from verbose output)
```text
$ logrotate -v /tmp/logrotate.conf 2>&1 | sed '/Creating new state/d'
reading config file /tmp/logrotate.conf
Reading state from file: /var/lib/logrotate/status
Allocating hash table for state file, size 1002 entries

Handling 1 logs

rotating pattern: /tmp/path/**/*
 1 bytes (4 rotations)
empty log files are rotated, old logs are removed
considering log /tmp/path/subA/log
  Now: 2022-11-17 12:34
  Last rotated at 2022-11-17 12:00
  log needs rotating
rotating log /tmp/path/subA/log, log->rotateCount is 4
dateext suffix '-20221117'
glob pattern '-[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]'
renaming /tmp/path/subA/log.4 to /tmp/path/subA/log.5 (rotatecount 4, logstart 1, i 4),
old log /tmp/path/subA/log.4 does not exist
renaming /tmp/path/subA/log.3 to /tmp/path/subA/log.4 (rotatecount 4, logstart 1, i 3),
old log /tmp/path/subA/log.3 does not exist
renaming /tmp/path/subA/log.2 to /tmp/path/subA/log.3 (rotatecount 4, logstart 1, i 2),
old log /tmp/path/subA/log.2 does not exist
renaming /tmp/path/subA/log.1 to /tmp/path/subA/log.2 (rotatecount 4, logstart 1, i 1),
old log /tmp/path/subA/log.1 does not exist
renaming /tmp/path/subA/log.0 to /tmp/path/subA/log.1 (rotatecount 4, logstart 1, i 0),
old log /tmp/path/subA/log.0 does not exist
log /tmp/path/subA/log.5 doesn't exist -- won't try to dispose of it
renaming /tmp/path/subA/log to /tmp/path/subA/log.1
creating new /tmp/path/subA/log mode = 0644 uid = 0 gid = 0
```

5. review `/tmp/path`
```text
$ find /tmp/path -type f -printf '%s\t%p\n' | sort -k2
512     /tmp/path/log
0       /tmp/path/subA/log
512     /tmp/path/subA/log.1
```

##### Second pass

Here is where a bug occurs.

6. fill dummy log files with data
```text
dd if=/dev/zero of=/tmp/path/subA/log count=1
```
7. review `/tmp/path` contents
```text
find /tmp/path -type f -printf '%s\t%p\n' | sort -k2
512     /tmp/path/log
512     /tmp/path/subA/log
512     /tmp/path/subA/log.1
````
8. run `logrotate` for second pass
```text
$ logrotate -v /tmp/logrotate.conf 2>&1 | sed '/Creating new state/d'
reading config file /tmp/logrotate.conf
Reading state from file: /var/lib/logrotate/status
Allocating hash table for state file, size 1002 entries

Handling 1 logs

rotating pattern: /tmp/path/**/*
 1 bytes (4 rotations)
empty log files are rotated, old logs are removed
considering log /tmp/path/subA/log
  Now: 2022-11-17 12:37
  Last rotated at 2022-11-17 12:34
  log needs rotating
considering log /tmp/path/subA/log.1
  Now: 2022-11-17 12:37
  Last rotated at 2022-11-17 12:00
  log needs rotating
rotating log /tmp/path/subA/log, log->rotateCount is 4
dateext suffix '-20221117'
glob pattern '-[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]'
renaming /tmp/path/subA/log.4 to /tmp/path/subA/log.5 (rotatecount 4, logstart 1, i 4),
old log /tmp/path/subA/log.4 does not exist
renaming /tmp/path/subA/log.3 to /tmp/path/subA/log.4 (rotatecount 4, logstart 1, i 3),
old log /tmp/path/subA/log.3 does not exist
renaming /tmp/path/subA/log.2 to /tmp/path/subA/log.3 (rotatecount 4, logstart 1, i 2),
old log /tmp/path/subA/log.2 does not exist
renaming /tmp/path/subA/log.1 to /tmp/path/subA/log.2 (rotatecount 4, logstart 1, i 1),
renaming /tmp/path/subA/log.0 to /tmp/path/subA/log.1 (rotatecount 4, logstart 1, i 0),
old log /tmp/path/subA/log.0 does not exist
log /tmp/path/subA/log.5 doesn't exist -- won't try to dispose of it
renaming /tmp/path/subA/log to /tmp/path/subA/log.1
creating new /tmp/path/subA/log mode = 0644 uid = 0 gid = 0
rotating log /tmp/path/subA/log.1, log->rotateCount is 4
dateext suffix '-20221117'
glob pattern '-[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]'
renaming /tmp/path/subA/log.1.4 to /tmp/path/subA/log.1.5 (rotatecount 4, logstart 1, i 4),
old log /tmp/path/subA/log.1.4 does not exist
renaming /tmp/path/subA/log.1.3 to /tmp/path/subA/log.1.4 (rotatecount 4, logstart 1, i 3),
old log /tmp/path/subA/log.1.3 does not exist
renaming /tmp/path/subA/log.1.2 to /tmp/path/subA/log.1.3 (rotatecount 4, logstart 1, i 2),
old log /tmp/path/subA/log.1.2 does not exist
renaming /tmp/path/subA/log.1.1 to /tmp/path/subA/log.1.2 (rotatecount 4, logstart 1, i 1),
old log /tmp/path/subA/log.1.1 does not exist
renaming /tmp/path/subA/log.1.0 to /tmp/path/subA/log.1.1 (rotatecount 4, logstart 1, i 0),
old log /tmp/path/subA/log.1.0 does not exist
log /tmp/path/subA/log.1.5 doesn't exist -- won't try to dispose of it
renaming /tmp/path/subA/log.1 to /tmp/path/subA/log.1.1
creating new /tmp/path/subA/log.1 mode = 0644 uid = 0 gid = 0
```
9. review `/tmp/path` contents
```text
$ find /tmp/path -type f -printf '%s\t%p\n' | sort -k2
512     /tmp/path/log
0       /tmp/path/subA/log
0       /tmp/path/subA/log.1
512     /tmp/path/subA/log.1.1
512     /tmp/path/subA/log.2
```
I expected
```text
$ find /tmp/path -type f -printf '%s\t%p\n' | sort -k2
0       /tmp/path/log
512     /tmp/path/log.1
0       /tmp/path/subA/log
512     /tmp/path/subA/log.1
512     /tmp/path/subA/log.2
```
(there are two problems; first the odd renaming of `/tmp/path/subA/log.1.1`, second skipping file `/tmp/path/log`)

##### Third Pass

10. fill dummy log file with data
```text
dd if=/dev/zero of=/tmp/path/subA/log count=1
```
11. review `/tmp/path` contents
```text
$ find /tmp/path -type f -printf '%s\t%p\n' | sort -k2
512     /tmp/path/log
512     /tmp/path/subA/log
0       /tmp/path/subA/log.1
512     /tmp/path/subA/log.1.1
512     /tmp/path/subA/log.2
```
12. run `logrotate` for third pass
```text
reading config file /tmp/logrotate.conf
Reading state from file: /var/lib/logrotate/status
Allocating hash table for state file, size 1002 entries

Handling 1 logs

rotating pattern: /tmp/path/**/*
 1 bytes (4 rotations)
empty log files are rotated, old logs are removed
considering log /tmp/path/subA/log
  Now: 2022-11-17 12:39
  Last rotated at 2022-11-17 12:37
  log needs rotating
considering log /tmp/path/subA/log.1
  Now: 2022-11-17 12:39
  Last rotated at 2022-11-17 12:37
  log does not need rotating (log size is below the 'size' threshold)
considering log /tmp/path/subA/log.1.1
  Now: 2022-11-17 12:39
  Last rotated at 2022-11-17 12:00
  log needs rotating
considering log /tmp/path/subA/log.2
  Now: 2022-11-17 12:39
  Last rotated at 2022-11-17 12:00
  log needs rotating
rotating log /tmp/path/subA/log, log->rotateCount is 4
dateext suffix '-20221117'
glob pattern '-[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]'
renaming /tmp/path/subA/log.4 to /tmp/path/subA/log.5 (rotatecount 4, logstart 1, i 4),
old log /tmp/path/subA/log.4 does not exist
renaming /tmp/path/subA/log.3 to /tmp/path/subA/log.4 (rotatecount 4, logstart 1, i 3),
old log /tmp/path/subA/log.3 does not exist
renaming /tmp/path/subA/log.2 to /tmp/path/subA/log.3 (rotatecount 4, logstart 1, i 2),
renaming /tmp/path/subA/log.1 to /tmp/path/subA/log.2 (rotatecount 4, logstart 1, i 1),
renaming /tmp/path/subA/log.0 to /tmp/path/subA/log.1 (rotatecount 4, logstart 1, i 0),
old log /tmp/path/subA/log.0 does not exist
log /tmp/path/subA/log.5 doesn't exist -- won't try to dispose of it
renaming /tmp/path/subA/log to /tmp/path/subA/log.1
creating new /tmp/path/subA/log mode = 0644 uid = 0 gid = 0
rotating log /tmp/path/subA/log.1.1, log->rotateCount is 4
dateext suffix '-20221117'
glob pattern '-[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]'
renaming /tmp/path/subA/log.1.1.4 to /tmp/path/subA/log.1.1.5 (rotatecount 4, logstart 1, i 4),
old log /tmp/path/subA/log.1.1.4 does not exist
renaming /tmp/path/subA/log.1.1.3 to /tmp/path/subA/log.1.1.4 (rotatecount 4, logstart 1, i 3),
old log /tmp/path/subA/log.1.1.3 does not exist
renaming /tmp/path/subA/log.1.1.2 to /tmp/path/subA/log.1.1.3 (rotatecount 4, logstart 1, i 2),
old log /tmp/path/subA/log.1.1.2 does not exist
renaming /tmp/path/subA/log.1.1.1 to /tmp/path/subA/log.1.1.2 (rotatecount 4, logstart 1, i 1),
old log /tmp/path/subA/log.1.1.1 does not exist
renaming /tmp/path/subA/log.1.1.0 to /tmp/path/subA/log.1.1.1 (rotatecount 4, logstart 1, i 0),
old log /tmp/path/subA/log.1.1.0 does not exist
log /tmp/path/subA/log.1.1.5 doesn't exist -- won't try to dispose of it
renaming /tmp/path/subA/log.1.1 to /tmp/path/subA/log.1.1.1
creating new /tmp/path/subA/log.1.1 mode = 0644 uid = 0 gid = 0
rotating log /tmp/path/subA/log.2, log->rotateCount is 4
dateext suffix '-20221117'
glob pattern '-[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]'
renaming /tmp/path/subA/log.2.4 to /tmp/path/subA/log.2.5 (rotatecount 4, logstart 1, i 4),
old log /tmp/path/subA/log.2.4 does not exist
renaming /tmp/path/subA/log.2.3 to /tmp/path/subA/log.2.4 (rotatecount 4, logstart 1, i 3),
old log /tmp/path/subA/log.2.3 does not exist
renaming /tmp/path/subA/log.2.2 to /tmp/path/subA/log.2.3 (rotatecount 4, logstart 1, i 2),
old log /tmp/path/subA/log.2.2 does not exist
renaming /tmp/path/subA/log.2.1 to /tmp/path/subA/log.2.2 (rotatecount 4, logstart 1, i 1),
old log /tmp/path/subA/log.2.1 does not exist
renaming /tmp/path/subA/log.2.0 to /tmp/path/subA/log.2.1 (rotatecount 4, logstart 1, i 0),
old log /tmp/path/subA/log.2.0 does not exist
log /tmp/path/subA/log.2.5 doesn't exist -- won't try to dispose of it
renaming /tmp/path/subA/log.2 to /tmp/path/subA/log.2.1
creating new /tmp/path/subA/log.2 mode = 0644 uid = 0 gid = 0
```
13. review `/tmp/path` contents
```text
$ find /tmp/path -type f -printf '%s\t%p\n' | sort -k2
512     /tmp/path/log
0       /tmp/path/subA/log
512     /tmp/path/subA/log.1
0       /tmp/path/subA/log.1.1
512     /tmp/path/subA/log.1.1.1
0       /tmp/path/subA/log.2
0       /tmp/path/subA/log.2.1
512     /tmp/path/subA/log.3
```
I expected log files like
```text
$ find /tmp/path -type f -printf '%s\t%p\n' | sort -k2
0       /tmp/path/log
512     /tmp/path/log.1
512     /tmp/path/log.2
0       /tmp/path/subA/log
512     /tmp/path/subA/log.1
512     /tmp/path/subA/log.2
512     /tmp/path/subA/log.3
```

<br/>

##### Using directive `olddir`

The [man page reads](https://linux.die.net/man/8/logrotate):

> Please use wildcards with caution. If you specify *, logrotate will rotate all files, including previously rotated ones. A way around this is to use the olddir directive or a more exact wildcard (such as *.log). 

Using the `olddir` directive does prevent the odd renaming. However, _**`olddir` loses the sub-directory path, so same-named files are ambiguously overwritten**_

1. Given `/tmp/logrotate.conf` contents
```text
create

/tmp/path/*/*
{
        rotate 4
        size 1
        olddir /tmp/old/
}
```

2. files
```text
$ find /tmp/path /tmp/old -type f -printf '%s\t%p\n' | sort -k2
512     /tmp/path/log
512     /tmp/path/subA/log
512     /tmp/path/subB/log
```

3. After the first pass of `logrotate`, results in ambiguous files
```text
$ find /tmp/path /tmp/old -type f -printf '%s\t%p\n' | sort -k2
512     /tmp/old/log.1
512     /tmp/old/log.2
512     /tmp/path/log
0       /tmp/path/subA/log
0       /tmp/path/subB/log
```

4. Fill the dummy files with content
```text
512     /tmp/old/log.1
512     /tmp/old/log.2
512     /tmp/path/log
512     /tmp/path/subA/log
512     /tmp/path/subB/log
```

5. After the second pass of `logrotate`, results in ambiguous and lost files
```text
$ find /tmp/path /tmp/old -type f -printf '%s\t%p\n' | sort -k2
512     /tmp/old/log.1
512     /tmp/old/log.2
512     /tmp/old/log.3
512     /tmp/path/log
0       /tmp/path/subA/log
0       /tmp/path/subB/log
```

##### Using directive `tabooext`

The [man page reads](https://linux.die.net/man/8/logrotate):

> tabooext [+] list
>    The current taboo extension list is changed (see the include directive for information on the taboo extensions). If a + precedes the list of extensions, the current taboo extension list is augmented, otherwise it is replaced. At startup, the taboo extension list contains .rpmorig, .rpmsave, ,v, .swp, .rpmnew, ~, .cfsaved and .rhn-cfg-tmp-*. 

Using the `tabooext` directive does not prevent the odd renaming. So given `/tmp/logrotate.conf` contents
```text
create
tabooext +.1
tabooext +.2
tabooext +.3

/tmp/path/*/*
{
        rotate 3
        size 1
}
```

After the second pass results in files
```text
$ find /tmp/path -type f -printf '%s\t%p\n' | sort -k2
512     /tmp/path/log
0       /tmp/path/subA/log
0       /tmp/path/subA/log.1
512     /tmp/path/subA/log.1.1
512     /tmp/path/subA/log.2
```

##### Recommendations

In my case, the directories and files that need to be rotated are very numerous. So it would be tedious to create a path declaration for each one. I imagine this is a common situation.

###### New directive `filter` or `nomatch`

The path declaration globbing does allow for matching, but there is not way to "unmatch" or "filter" files. Currently, there is no logrotate mechanism to allow for this.

I recommend adding a directive `filter` to not attempt rotation of certain file patterns.
So `logrotate.conf` contents could be:

```text
create

/tmp/path/*/*
{
        rotate 3
        size 1
        filter *.1
        filter *.2
        filter *.3
}
```
`logrotate` would not rotate files with those names (or perhaps directive name `nomatch`).

###### Modify `olddir` to keep subpath

Modifying `olddir` to include the sub-directory path would keep relevant information and prevent file overwrites.

So in the example above, after the first pass there would be files
```text
$ find /tmp/path /tmp/old -type f -printf '%s\t%p\n' | sort -k2
512     /tmp/old/subA/log.1
512     /tmp/old/subB/log.1
512     /tmp/path/log
0       /tmp/path/subA/log
0       /tmp/path/subB/log
```

---

Using logrotate version `3.18.0`.

This Issue addresses two or more problems and may need to be split.

**Comment by kdudka** [comment]
I am afraid this is the expected behavior, already documented in the man page:

> Please use wildcards with caution.  If you specify *, logrotate will rotate all files, including previously rotated ones.
> A way around this is to use the olddir directive or a more exact wildcard (such as *.log).
